core/event.go

```diff
@@ -57,3 +62,32 @@ func (n *nullEvent) Type() string {
 
 // Sanity check
 var _ Event = (*nullEvent)(nil)
+
+type ActionType int
+
+const (
+	Create ActionType = iota
+	Save
+	Delete
+)
+
+// Action is a operation done in the model
+type Action struct {
+	// Type of the action
+	Type ActionType
+	// EntityID of the instance in action
+	EntityID EntityID
+	// EntityType of the instance in action
+	EntityType string
+	// Previous is the instance before the action
+	Previous interface{}
```

Comment:
`Previous` is needed by the `EventCreator`, for creating deltas of state.
Example: the current json-patcher uses `Previous` and `Current` in update actions to calculate the corresponding json-patch which will be the payload of the event that generates.

Now that you're mentioning this, I think can be avoided, since the `EventCreator` has access to the underlying datastore and can get the "previous" value with the `EntityID` also stored in the `Action`. (since the datastore is still unchanged at this point). Maybe you were thinking that.

Reply:
Ah, ok. Makes sense. I wasn't thinking that... just needed a refresher